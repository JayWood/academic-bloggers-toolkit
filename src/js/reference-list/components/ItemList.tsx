import { action, IObservableArray, IObservableObject, IObservableValue, ObservableMap } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';

import { Card } from './Card';

interface UI {
    cited: {
        readonly maxHeight: string;
        isOpen: IObservableValue<boolean>;
    } & IObservableObject;
    uncited: {
        readonly maxHeight: string;
        isOpen: IObservableValue<boolean>;
    } & IObservableObject;
    readonly [k: string]: any;
}

interface Props {
    readonly items: CSL.Data[];
    readonly id: 'cited' | 'uncited';
    readonly children: string;
    ui: UI;
    selectedItems: IObservableArray<string>;
    CSL: ObservableMap<CSL.Data>;
    onEditReference(referenceId: string): void;
}

@observer
export class ItemList extends React.PureComponent<Props> {
    @action
    singleClick = () => {
        this.props.ui[this.props.id].isOpen.set(!this.props.ui[this.props.id].isOpen.get());
    };

    @action
    doubleClick = () => {
        this.props.ui.cited.isOpen.set(false);
        this.props.ui.uncited.isOpen.set(false);
        this.props.ui[this.props.id].isOpen.set(true);
    };

    @action
    toggleSelect = (e: React.MouseEvent<HTMLDivElement>) => {
        this.props.selectedItems.remove(e.currentTarget.id) ||
            this.props.selectedItems.push(e.currentTarget.id);
    };

    render() {
        const { CSL, children, id, items, onEditReference, selectedItems, ui } = this.props;
        if (!items || items.length === 0) return null;
        return (
            <div>
                <div
                    className="abt-item-heading"
                    role="menubar"
                    onClick={this.singleClick}
                    onDoubleClick={this.doubleClick}
                >
                    <div className="abt-item-heading__label" children={children} />
                    <div className="abt-item-heading__badge" children={items.length} />
                </div>
                {ui[id].isOpen.get() &&
                    <Items
                        items={items}
                        selectedItems={selectedItems}
                        withTooltip={id === 'cited'}
                        CSL={CSL}
                        onEditReference={onEditReference}
                        id={id}
                        style={{ maxHeight: ui.maxHeight }}
                        onClick={this.toggleSelect}
                    />}
            </div>
        );
    }
}

interface ItemsProps extends React.HTMLProps<HTMLElement> {
    readonly items: CSL.Data[];
    readonly selectedItems: string[];
    readonly withTooltip: boolean;
    CSL: ObservableMap<CSL.Data>;
    onEditReference(referenceId: string): void;
}

@observer
class Items extends React.Component<ItemsProps, {}> {
    handleScroll = (e: React.WheelEvent<HTMLDivElement>) => {
        e.stopPropagation();
        const atTopAndScrollingUp: boolean = e.currentTarget.scrollTop === 0 && e.deltaY < 0;
        const atBottomAndScollingDown: boolean =
            Math.floor(e.currentTarget.scrollTop + e.currentTarget.offsetHeight) ===
                e.currentTarget.scrollHeight && e.deltaY > 0;
        if (atTopAndScrollingUp || atBottomAndScollingDown) {
            e.preventDefault();
        }
    };

    @action
    editSingleReference = (e: React.MouseEvent<HTMLDivElement>) => {
        const referenceId = e.currentTarget.id;
        this.props.onEditReference(referenceId);
    };

    render() {
        return (
            <div
                onWheel={this.handleScroll}
                id={this.props.id}
                className="abt-items"
                style={{
                    ...this.props.style,
                    overflowX: 'hidden',
                    overflowY: 'auto',
                    transition: '.2s',
                }}
            >
                {this.props.items.map((r, i) =>
                    <Card
                        key={r.id}
                        id={r.id}
                        CSL={r}
                        onClick={this.props.onClick}
                        onDoubleClick={this.editSingleReference}
                        index={`${i + 1}`}
                        isSelected={this.props.selectedItems.indexOf(r.id!) > -1}
                        showTooltip={this.props.withTooltip}
                    />
                )}
            </div>
        );
    }
}
