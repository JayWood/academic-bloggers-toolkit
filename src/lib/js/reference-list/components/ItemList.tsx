import * as React from 'react';
import { Card } from './Card';
import { observer } from 'mobx-react';
import { preventScrollPropagation } from '../../utils/HelperFunctions';

interface Props extends React.HTMLProps<HTMLElement> {
    readonly items: CSL.Data[];
    readonly selectedItems: string[];
    click: (id: string, isSelected: boolean) => void;
    toggle: (id: string, explode?: boolean) => void;
    isOpen: boolean;
    maxHeight: string;
}

@observer
export class ItemList extends React.PureComponent<Props, {}> {

    singleClick = () => {
        this.props.toggle(this.props.id);
    }

    doubleClick = () => {
        this.props.toggle(this.props.id, true);
    }

    render() {
        const { items, selectedItems, click, children, isOpen, maxHeight, id } = this.props;
        if (!items) return null;
        return (
            <div>
                <div
                    className="abt-item-heading"
                    role="menubar"
                    onClick={this.singleClick}
                    onDoubleClick={this.doubleClick}
                >
                    <div className="abt-item-heading__label" children={children} />
                    <div className="abt-item-heading__badge" children={items.length} />
                </div>
                { isOpen &&
                    <Items
                        id={id}
                        style={{maxHeight}}
                        items={items}
                        selectedItems={selectedItems}
                        click={click}
                    />
                }
            </div>
        );
    }
}

interface ItemsProps extends React.HTMLProps<HTMLElement> {
    readonly items: CSL.Data[];
    readonly selectedItems: string[];
    readonly click: (id: string, isSelected: boolean) => void;
}

@observer
class Items extends React.Component<ItemsProps, {}> {

    element: HTMLDivElement;
    handleScroll = preventScrollPropagation.bind(this);

    bindRefs = (c: HTMLDivElement) => {
        this.element = c;
    }

    render() {
        return (
            <div
                ref={this.bindRefs}
                onWheel={this.handleScroll}
                id={this.props.id}
                className="abt-items"
                style={Object.assign({}, this.props.style, {
                    overflowX: 'hidden',
                    overflowY: 'auto',
                    transition: '.2s',
                })}
            >
                {
                    this.props.items.map(r =>
                        <Card
                            key={r.id}
                            CSL={r}
                            isSelected={this.props.selectedItems.indexOf(r.id) > -1}
                            id={r.id}
                            click={this.props.click}
                        />
                    )
                }
            </div>
        );
    }
}
