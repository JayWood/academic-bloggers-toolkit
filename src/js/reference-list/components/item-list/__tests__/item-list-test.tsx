import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { observable } from 'mobx';
import * as React from 'react';

import UIStore from 'stores/ui/reference-list';
import ItemList from '..';

const createItems = (numItems: number) => {
    if (numItems === -1) {
        return {
            CSL: [],
            items: [],
        };
    }
    const items = Array(numItems)
        .fill(0)
        .map(
            (_, i) =>
                ({
                    id: `${i + 1}`,
                    type: 'article-journal',
                    title: `Test item ${i + 1}`,
                } as CSL.Data),
        );
    return {
        items,
        CSL: items.map(item => [item.id, item]),
    };
};
const CSLMap = observable.map<CSL.Data>();
const onEditReference = jest.fn();

const setup = (numItems = 3, selectedNum: string[] = []) => {
    const { CSL, items } = createItems(numItems);
    const ui = new UIStore(false);
    ui.selected.replace(selectedNum);
    CSLMap.replace(CSL);
    const component = shallow(
        <ItemList items={items} id="cited" ui={ui} CSL={CSLMap} onEditReference={onEditReference}>
            Hello World
        </ItemList>,
    );
    return {
        component,
        instance: component.instance() as ItemList,
        ui,
    };
};

describe('<ItemList />', () => {
    it('should match snapshot', () => {
        let { component } = setup();
        expect(toJSON(component)).toMatchSnapshot();

        ({ component } = setup(3, ['1']));
        expect(toJSON(component)).toMatchSnapshot();

        ({ component } = setup(-1));
        expect(component.isEmptyRender()).toBe(true);

        ({ component } = setup(0));
        expect(component.isEmptyRender()).toBe(true);
    });
    it('should handle click', () => {
        const { component, ui } = setup();
        expect(ui.cited.isOpen).toBeTruthy();
        component.find('.heading').simulate('click');
        expect(ui.cited.isOpen).toBeFalsy();
    });
    it('should handle double click', () => {
        const { component, ui } = setup();
        ui.uncited.isOpen = true;
        expect(ui.cited.isOpen).toBeTruthy();
        expect(ui.uncited.isOpen).toBeTruthy();
        component.find('.heading').simulate('doubleClick');
        expect(ui.cited.isOpen).toBeTruthy();
        expect(ui.uncited.isOpen).toBeFalsy();
    });
    it('should handle item click', () => {
        const { component, ui } = setup(3, ['2']);
        const items = component.find('Item').first();
        expect(ui.selected.slice()).toEqual(['2']);

        items.simulate('click', { currentTarget: { id: '1' } });
        expect(ui.selected.slice()).toEqual(['2', '1']);

        items.simulate('click', { currentTarget: { id: '2' } });
        expect(ui.selected.slice()).toEqual(['1']);
    });
    it('should handle shift-click selection', () => {
        const { component, ui } = setup(5, ['1']);
        const items = component.find('Item').at(4);
        expect(ui.selected.slice()).toEqual(['1']);

        items.simulate('click', { currentTarget: { id: '5' }, shiftKey: true });
        expect(ui.selected.slice()).toEqual(['1', '2', '3', '4', '5']);
    });
    it('should pass over previously selected items with shift-click', () => {
        const { component, ui } = setup(10, ['3', '5', '7', '9']);
        const items = component.find('Item').first();
        expect(ui.selected.slice()).toEqual(['3', '5', '7', '9']);

        items.simulate('click', { currentTarget: { id: '1' }, shiftKey: true });
        expect(ui.selected.slice().sort()).toEqual(['1', '2', '3', '4', '5', '6', '7', '8', '9']);
    });
});

// FIXME:
// describe('<Items />', () => {
//     const setup = setupItems;
//     it('should match snapshot', () => {
//         let { component } = setup();
//         expect(toJSON(component)).toMatchSnapshot();

//         ({ component } = setup(3, ['1']));
//         expect(toJSON(component)).toMatchSnapshot();
//     });
//     it('should trigger edit reference on double click', () => {
//         const { component } = setup();
//         expect(onEditReference).not.toHaveBeenCalled();
//         component
//             .find('Card')
//             .first()
//             .simulate('doubleClick', { currentTarget: { id: '2' } });
//         expect(onEditReference).toHaveBeenCalled();
//         expect(onEditReference).toBeCalledWith('2');
//     });
//     describe('scroll handler tests', () => {
//         const stopPropagation = jest.fn();
//         const preventDefault = jest.fn();
//         function makeEvent({
//             deltaY = 0,
//             offsetHeight = 250,
//             scrollHeight = 500,
//             scrollTop = 0,
//         }: { [k: string]: number } = {}) {
//             return {
//                 deltaY,
//                 currentTarget: {
//                     offsetHeight,
//                     scrollHeight,
//                     scrollTop,
//                 },
//                 preventDefault,
//                 stopPropagation,
//             };
//         }
//         beforeEach(() => jest.resetAllMocks());
//         test('at top & scrolling up', () => {
//             const { component } = setup();
//             component.simulate('wheel', makeEvent({ deltaY: -5 }));
//             expect(stopPropagation).toHaveBeenCalledTimes(1);
//             expect(preventDefault).toHaveBeenCalledTimes(1);
//         });
//         test('at bottom & scrolling down', () => {
//             const { component } = setup();
//             component.simulate('wheel', makeEvent({ deltaY: 5, scrollTop: 250 }));
//             expect(stopPropagation).toHaveBeenCalledTimes(1);
//             expect(preventDefault).toHaveBeenCalledTimes(1);
//         });
//         test('at bottom & scrolling up', () => {
//             const { component } = setup();
//             component.simulate('wheel', makeEvent({ deltaY: -5, scrollTop: 250 }));
//             expect(stopPropagation).toHaveBeenCalledTimes(1);
//             expect(preventDefault).not.toHaveBeenCalled();
//         });
//         test('at top & scrolling down', () => {
//             const { component } = setup();
//             component.simulate('wheel', makeEvent({ deltaY: 5 }));
//             expect(stopPropagation).toHaveBeenCalledTimes(1);
//             expect(preventDefault).not.toHaveBeenCalled();
//         });
//         test('in middle & scrolling', () => {
//             const { component } = setup();
//             component.simulate('wheel', makeEvent({ deltaY: -5, scrollTop: 50 }));
//             component.simulate('wheel', makeEvent({ deltaY: 5, scrollTop: 50 }));
//             expect(stopPropagation).toHaveBeenCalledTimes(2);
//             expect(preventDefault).not.toHaveBeenCalled();
//         });
//     });
// });
