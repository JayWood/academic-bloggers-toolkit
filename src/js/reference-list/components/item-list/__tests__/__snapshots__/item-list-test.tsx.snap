// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemList /> should match snapshot 1`] = `
<div>
  <div
    className="heading"
    onClick={[Function]}
    onDoubleClick={[Function]}
    role="menubar"
  >
    <span>
      Hello World
    </span>
    <Badge
      count={3}
    />
  </div>
  <div
    className="abt-items"
    id="cited"
    onWheel={[Function]}
    style={
      Object {
        "maxHeight": "400px",
        "overflowX": "hidden",
        "overflowY": "auto",
        "transition": ".2s",
      }
    }
  >
    <Item
      CSL={
        Object {
          "id": "1",
          "title": "Test item 1",
          "type": "article-journal",
        }
      }
      id="1"
      index={1}
      indexOnHover={true}
      isSelected={false}
      key="1"
      onClick={[Function]}
      onDoubleClick={[Function]}
    />
    <Item
      CSL={
        Object {
          "id": "2",
          "title": "Test item 2",
          "type": "article-journal",
        }
      }
      id="2"
      index={2}
      indexOnHover={true}
      isSelected={false}
      key="2"
      onClick={[Function]}
      onDoubleClick={[Function]}
    />
    <Item
      CSL={
        Object {
          "id": "3",
          "title": "Test item 3",
          "type": "article-journal",
        }
      }
      id="3"
      index={3}
      indexOnHover={true}
      isSelected={false}
      key="3"
      onClick={[Function]}
      onDoubleClick={[Function]}
    />
  </div>
</div>
`;

exports[`<ItemList /> should match snapshot 2`] = `
<div>
  <div
    className="heading"
    onClick={[Function]}
    onDoubleClick={[Function]}
    role="menubar"
  >
    <span>
      Hello World
    </span>
    <Badge
      count={3}
    />
  </div>
  <div
    className="abt-items"
    id="cited"
    onWheel={[Function]}
    style={
      Object {
        "maxHeight": "400px",
        "overflowX": "hidden",
        "overflowY": "auto",
        "transition": ".2s",
      }
    }
  >
    <Item
      CSL={
        Object {
          "id": "1",
          "title": "Test item 1",
          "type": "article-journal",
        }
      }
      id="1"
      index={1}
      indexOnHover={true}
      isSelected={true}
      key="1"
      onClick={[Function]}
      onDoubleClick={[Function]}
    />
    <Item
      CSL={
        Object {
          "id": "2",
          "title": "Test item 2",
          "type": "article-journal",
        }
      }
      id="2"
      index={2}
      indexOnHover={true}
      isSelected={false}
      key="2"
      onClick={[Function]}
      onDoubleClick={[Function]}
    />
    <Item
      CSL={
        Object {
          "id": "3",
          "title": "Test item 3",
          "type": "article-journal",
        }
      }
      id="3"
      index={3}
      indexOnHover={true}
      isSelected={false}
      key="3"
      onClick={[Function]}
      onDoubleClick={[Function]}
    />
  </div>
</div>
`;
